src/dd.ts

```diff
@@ -0,0 +1,156 @@
+        }
+
+        return this.ddQueueContract;
+    }
+
+    public async getFee(): Promise<bigint> {
+        const queue = await this.getQueueContract();
+
+        let fee = this.cachedFee;
+        if (!fee || fee.timestamp + DD_FEE_LIFETIME * 1000 < Date.now()) {
+            const queue = await this.getQueueContract();
+            const fetchedFee = await this.network.getDirectDepositFee(queue);
+            fee = {fee: fetchedFee, timestamp: Date.now()};
+            this.cachedFee = fee;
+        }
+
+        return fee.fee;
+    }
+
+    public async prepareDirectDeposit(
+        type: DirectDepositType,
+        zkAddress: string,
+        amount: bigint, // in native resolution (wei for DirectDepositType.Native)
+        fallbackAddress: string,
+        feeAlreadyIncluded: boolean = true, // calculate and add required fee amount when false
+    ): Promise<PreparedTransaction> {
+        if (type == DirectDepositType.Native && !this.isNativeSupported) {
+            throw new InternalError(`Native direct deposits are not supported in this pool`);
+        }
+
+        const queue = await this.getQueueContract();
+
+        let addedFee = 0n;
+        if (!feeAlreadyIncluded) {
+            addedFee = await this.getFee();
+        }
+
+        switch (type) {
+            case DirectDepositType.Token:
+                return this.network.createDirectDepositTx(queue, amount + addedFee, zkAddress, fallbackAddress);
+                
+            case DirectDepositType.Native:
+                return this.network.createNativeDirectDepositTx(queue, amount + addedFee, zkAddress, fallbackAddress);       
+
+            default:
+                throw new InternalError(`Unsupported direct deposit type ${type}`);
+        }
+    }
+
+    public async pendingDirectDeposits(): Promise<DirectDeposit[]> {
+        if (this.subgraphName == 'zkbob-bob-goerli') {
+            const allPendingDDs = (await execute(ddQuery, {})).data.directDeposits;
+            if (Array.isArray(allPendingDDs)) {
+                const myPendingDDs: DirectDeposit[] = await allPendingDDs.reduce(async (result, dd) => {
+                    const d = BigInt(dd.zkAddress_diversifier);
+                    const p_d = BigInt(dd.zkAddress_pk);
+                    const zkAddress = await this.state.assembleAddress(d.toString(), p_d.toString());
+                    const isOwn = await this.state.isOwnAddress(zkAddress);
+                    if (isOwn) {
+                        const ownDD: DirectDeposit =  {
+                            id: BigInt(dd.id),
+                            state: DirectDepositState.Queued,
+                            amount: BigInt(dd.deposit),
+                            destination: zkAddress,
+                            fallback: dd.fallbackUser,
+                            timestamp: Number(dd.tsInit),
+                            queueTxHash: dd.txInit,
+                        };
+                        (await result).push(ownDD);
```

Comment:
Why do we need await here?
```suggestion
                        result.push(ownDD);
```

Reply:
It's because the result is a `Promise` due to reduce callback function is asynchronous.  Invoking of `push` method on `Promise` will cause an exception
<img width="779" alt="image" src="https://github.com/zkBob/zkbob-client-js/assets/1415489/5727e86e-76ca-4c9d-9bb7-2339eec511ca">

Reply:
`result` will be an array during the first iteration, but a `Promise` in later iterations. I personally don't like an idea of creating a recursive promise here. It works, but it looks really strange to me, I think it's better if we use a more common `Promise.all` + `map` + `filter` approach.

Reply:
Fixed in https://github.com/zkBob/zkbob-client-js/pull/144/commits/fe16410083cfe48924cb15b6c6201a3902ffc019